frontend/src/components/Tree/Tree.jsx

```diff
@@ -0,0 +1,138 @@
+import React, { useState, useRef, useEffect } from "react";
+import styles from "./Tree.module.css";
+import { select, hierarchy, tree, linkHorizontal } from "d3";
+import ResizeObserver from "resize-observer-polyfill";
+import "./Tree.module.css";
+
+const useResizeObserver = (ref) => {
+  const [dimensions, setDimensions] = useState(null);
+  useEffect(() => {
+    const observeTarget = ref.current;
+    const resizeObserver = new ResizeObserver((entries) => {
+      entries.forEach((entry) => {
+        setDimensions(entry.contentRect);
+      });
+    });
+    resizeObserver.observe(observeTarget);
+    return () => {
+      resizeObserver.unobserve(observeTarget);
+    };
+  }, [ref]);
+  return dimensions;
+};
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+export function Tree({ jsonData, onNodeClick }) {
+  const [currentView, setCurrentView] = useState({});
+  const svgRef = useRef();
+  const wrapperRef = useRef();
+  const dimensions = useResizeObserver(wrapperRef);
+
+  // We save data to see if it changed
+  const previouslyRenderedData = usePrevious(currentView);
+
+  useEffect(() => {
+    setCurrentView(jsonData);
+  }, [jsonData]);
+
+  // Will be called initially and on every data change
+  useEffect(() => {
+    const svg = select(svgRef.current);
+
+    // Use dimensions from useResizeObserver,
+    // but use getBoundingClientRect on initial render
+    // (dimensions are null for the first render)
+    const { width, height } =
+      dimensions || wrapperRef.current.getBoundingClientRect();
```

Comment:
Do we need to pass the dimensions here when we can just use the wrapperRef's current dimensions?

Reply:
Thanks, found a way of avoiding this.